design-proposals/klip-43-timestamp-data-type-support.md

```diff
@@ -0,0 +1,159 @@
+```json
+{
+  "type": "long",
+  "logicalType": "timestamp-millis"
+}
+```
+The Avro deserializer KSQL uses supports this.
+
+#### JSON/Delimited
+
+Timestamps will get stored in JSON and CSV files as long values. The KSQL JSON and delimited deserializers
+will be updated to parse timestamps.
+
+### UDFs
+
+The following UDFs should be updated to use the TIMESTAMP type instead of BIGINT:
+
+* TIMESTAMPTOSTRING
+* STRINGTOTIMESTAMP
+* UNIX_TIMESTAMP
+
+The following UDFs should also be added to support conversions between TIMESTAMP and BIGINT representations of time:
+
+* TIMESTAMPTOBIGINT
+* BIGINTTOTIMESTAMP
+
+There are a few existing UDFs that deal with dates. These should be left as is until a DATE type is implemented:
+
+* UNIX_DATE
+* DATETOSTRING
+* STRINGTODATE
+
+### Casting/Converting
+
+Since the above UDFs already handle conversions, casting between TIMESTAMP and STRING or BIGINT is
+unnecessary. If DATE and TIME types are implemented, then the conversions between them and TIMESTAMP
+will be as follows:
+
+| From      | To        | Result                                                |
+|---------- |:----------|:------------------------------------------------------|
+| TIMESTAMP | DATE      | DATE type representing date portion of the TIMESTAMP. |
+| TIMESTAMP | TIME      | TIME type representing time portion of the TIMESTAMP. |
+| DATE      | TIMESTAMP | TIMESTAMP with time portion set to 00:00:00.          |
+| TIME      | TIMESTAMP | TIMESTAMP with date portion set to Unix epoch.        |
+
+### Arithmetic operations and comparisons
+
+At the very least, we would want KSQL to be able to compare, add and subtract TIMESTAMPS.
+
+[MySQL](https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html) and [MariaDB](https://mariadb.com/kb/en/date-time-functions/)
+have very complete lists of built-in arithmetic functions, while [PostgreSQL](https://www.postgresql.org/docs/9.0/functions-datetime.html)
+defines behaviors for arithmetic operators.
+
+Because functions are more well-defined, we should opt for built-in arithmetic functions
+instead of using operators. The following functions are necessary:
+
+* ADDTIME(timestamp1, timestamp2)
+* SUBTIME(timestamp1, timestamp2)
+
+As for comparisons, the following expressions should be supported:
+```
+time_stamp1 < time_stamp2
+time_stamp1 > time_stamp2
+time_stamp1 = time_stamp2
+time_stamp1 BETWEEN time_stamp2 AND time_stamp3
+```
+
+Comparisons between TIMESTAMPS and other data types should not be allowed.
+
+### Window units
+
+It might make sense to give window units a timestamp representation so that they could be used as time intervals in TIMESTAMP arithmetic. For example,
+
+```roomsql
+SELECT ADDTIME(TIME, 1 DAY) FROM FOO;
+```
+
+Because windowing is handled separately from expression evaluation, doing this will not have any
+impact on windows.
+
+## Test plan
+
+There should be an integration test with Kafka Connect and Schema Registry.
```

Comment:
What about QTT tests with different serdes? and the supported UDFs?

Do we need a performance test plan for TIMESTAMP serdes? We have one for some serde formats, but I don't know if we need one for Timestamps (your call).

Reply:
I don't think having the TIMESTAMP type in a schema would have any effect on performance, and `java.time.Instant` does not seem to have any performance issues. It might be good to check at some point, but I don't think we will need to benchmark it.